src/main/java/org/derive4j/hkt/TypeEq.java

```diff
@@ -220,281 +341,91 @@ private TypeEq() {
   }
+}
 
-  /**
-   * Leibnizian equality states that two things are equal if you can substitute one for the other in all contexts.
-   *
-   * @param fa a term whose type last parameter is {@link A}.
-   * @param <f> type constructor witness of {@code fa} type.
-   * @return the input value with {@link A} substituted by {@link B} in its type.
-   */
-  public abstract <f> __<f, B> subst(__<f, A> fa);
-
-  /**
-   * If two things are equal you can convert one to the other (safe cast).
-   *
-   * @param a a value of type {@link A} that will be coerced into type {@link B}.
-   * @return the same value, after type coercion.
-   */
-  public final B coerce(A a) {
-    return Identity.ofHkt(subst(new Identity<>(a))).runIdentity;
-  }
-
-  /**
-   * Equality is transitive.
-   *
-   * @param that another TypeEq to compose with.
-   * @param <C> left operand of the transitive type equality.
-   * @return the composition of the TypeEq instances.
-   */
-  public final <C> TypeEq<C, B> compose(TypeEq<C, A> that) {
-    return new TypeEq<C, B>() {
-      @Override public <f> __<f, B> subst(__<f, C> fa) {
-        return TypeEq.this.subst(that.subst(fa));
-      }
-    };
-  }
-
-  /**
-   * Equality is transitive.
-   *
-   * @param that another TypeEq to be composed with.
-   * @param <C> right operand of the transitive type equality.
-   * @return the composition of the TypeEq instances.
-   */
-  public final <C> TypeEq<A, C> andThen(TypeEq<B, C> that) {
-    return that.compose(this);
-  }
-
-  /**
-   * Equality is symmetric.
-   *
-   * @return the type equality seen from the other side.
-   */
-  public final TypeEq<B, A> symm() {
-    return Symm.ofHkt(subst(new Symm<>(refl()))).typeEq;
-  }
-
-  /**
-   * The type equality can be lifted into any type constructor.
-   *
-   * @param <f> a type constructor witness.
-   * @return the type equality in the context of the specified type constructor.
-   */
-  public final <f> TypeEq<__<f, A>, __<f, B>> lift() {
-    return Lift.ofHkt(subst(new Lift<>(TypeEq.<__<f, A>>refl()))).unlift;
-  }
-
-  /**
-   * The type equality can be lifted into any type constructor, at any position.
-   *
-   * @param <f> a type constructor witness.
-   * @param <C> the last type variable (not substituted).
-   * @return the type equality in the context of the specified type constructor.
-   */
-  public final <f, C> TypeEq<__<__<f, A>, C>, __<__<f, B>, C>> lift2() {
-    return Lift2.ofHkt(subst(new Lift2<>(TypeEq.<__<__<f, A>, C>>refl()))).unlift2;
-  }
-
-  /**
-   * Type equalities can be lifted into any type constructor, at any positions.
-   *
-   * @param cd a type equality witness for last type variable of any type constructor.
-   * @param <C> the last type variable before substitution.
-   * @param <D> the last type variable after substitution.
-   * @return a factory to lift the TypeEq instances into any type constructor.
-   */
-  public final <C, D> Lift2TypeEq<A, C, B, D> lift2(TypeEq<C, D> cd) {
```

Comment:
I am a bit concerned about the removal of those type inference helpers (`Lift2TypeEq`, `Lift3TypeEq`). I think they can be useful. @gneuvill could you add them back in some way?